Theme selector

Lemonatix opened the pull request:
First attempt of adding a working theme-switch for the app.

Lemonatix committed:
Creating theme-switch (V1.0)

Lemonatix committed:
Create theme-switch-dependencies

adds dependencies for theme-switch (V1.0)

Lemonatix committed:
theme-switch (V1.0).dart

Renames file to with a format

Lemonatix committed:
Rename theme-switch (V1.0).dart

Lemonatix committed:
Rename theme-switch-dependences to theme-switch-dependences.yml

Adds file tag .yml

Lemonatix committed:
Update theme-switch (V1.0).dart

Fixed minor syntax mistakes

Lemonatix committed:
Adds folder for theme-switch

Lemonatix committed:
Adds dependencies to theme-switch folder

Lemonatix committed:
Adds main (V1.0) to theme-switch

biosfood commented on theme-switch/main-theme-switch (V1.0).dart:
```diff
@@ -0,0 +1,60 @@
+    runApp(MyApp());
+}
+
+// MyApp = stateful widget, that maintains current theme state
```
remove this, this seems clear

Lemonatix replied:
Is removed now

biosfood commented on theme-switch/main-theme-switch (V1.0).dart:
```diff
@@ -0,0 +1,60 @@
+    void initState() { // initState Method: loads the theme preference from shared preference, when App starts
+        super.initState();
+        _loadTheme();
+    }
```
spacing

biosfood commented on theme-switch/main-theme-switch (V1.0).dart:
```diff
@@ -0,0 +1,60 @@
+                        setState(() {
+                            _isDarkMode = value;
+                            _saveTheme(value);
+                        }
```
Indentation and semicolon, or remove brackets

biosfood commented on theme-switch/main-theme-switch (V1.0).dart:
```diff
@@ -0,0 +1,60 @@
+// theme-switch (1.0) as .dart file
```
in lib folder, maybe UI or settings folder
remove 1.0 from file name and comment
remove comment

biosfood commented on theme-switch/main-theme-switch (V1.0).dart:
```diff
@@ -0,0 +1,60 @@
+import 'package:flutter/material.dart';
+import 'package:shared_preferences/shared_preferences.dart'
+
+void main() {
```
no main function other than in main.dart or libs

Lemonatix replied:
main.dart is now consisting of the switch selector code

biosfood commented on theme-switch/theme-switch-dependences.yml:
```diff
@@ -0,0 +1,7 @@
+# Add Dependencies
+# shared_preferences package must be included in pubspec.yaml (file for saving user's preference)
+
+dependencies:
```
not in a separate file

biosfood commented on theme-switch/main-theme-switch (V1.0).dart:
```diff
@@ -0,0 +1,60 @@
+        _loadTheme();
+    }
+    // _loadTheme Method: asynchronously loads preference from SharedPreferences, when App starts
+    _loadTheme() async {
```
maybe just wait for this when booting?

Maybe profile this on some devices and only show the scene when everything is loaded

Lemonatix replied:
Still need to figure out, how to test it with apple device

biosfood commented on theme-switch/main-theme-switch (V1.0).dart:
```diff
@@ -0,0 +1,60 @@
+                    title: Text('Theme Switch'),
+            ),
+            body: Center(
+                child: SwitchListTile( // Allows user to toggle theme, which updates the state and saves the preference
```
Put this in a burger or side bar

Lemonatix replied:
Added Drawer (some kind of side bar)

biosfood commented on theme-switch/main-theme-switch (V1.0).dart:
```diff
@@ -0,0 +1,60 @@
+            _isDarkMode = (prefs.getBool('isDarkMode') ?? false);
+        })
+    }
+    // _saveTheme Method: asynchronously saves theme preference to when changed
```
spacing

biosfood commented on theme-switch/main-theme-switch (V1.0).dart:
```diff
@@ -0,0 +1,60 @@
+    }
+    // _saveTheme Method: asynchronously saves theme preference to when changed
+    _saveTheme(bool value) async {
+        _SharedPreferences prefs = await SharedPreferences.getInstance();
```
use a global SharedPreferences object
This lets us avoid waiting here again

Lemonatix replied:
Added a new ShapedPreferences object in main.dart

biosfood commented on theme-switch/theme-switch-dependences.yml:
```diff
@@ -0,0 +1,7 @@
+dependencies:
+  flutter:
+    sdk: flutter
+  shared_preference: ^2.0.13
```
Is this the right choice?
Some others maybe chime in here:
Is this enough for our storage needs or do we also need another storage?

Lemonatix replied:
Going with this for now

biosfood reviewed (changes requested):
Seems good in gerneral, I left some notes for the code style

Lemonatix committed:
Update and rename main-theme-switch (V1.0).dart to main-theme-switch.dart

removes some unnecessary comments and changes the file name

Lemonatix committed:
Update main-theme-switch.dart

Removes unnecessary comments, should improve spacing, adds a global SharendPreference object and adds something like a sidebar.

Testing still awaits and this will replace main.dart in lib

Lemonatix committed:
Update pubspec.yaml

Adds shared_preference to this file instead of having a seperate one

Lemonatix committed:
Delete theme-switch/theme-switch-dependences.yml

Lemonatix committed:
Update main-theme-switch.dart

Minor changes to make the code ready to replace current main.dart

Lemonatix committed:
Update main.dart

New main.dart with included theme selector

Lemonatix committed:
Delete theme-switch directory

biosfood commented on lib/main.dart:
```diff
@@ -1,20 +1,80 @@
+    void _loadTheme() {
+        setState(() {
+            _isDarkMode = (prefs?.getBool('isDarkMode') ?? false);
+    });
```
still an indentation problem here

biosfood commented on lib/main.dart:
```diff
@@ -1,20 +1,80 @@
 
-void main() {
-  runApp(const MainApp());
+SharedPreferences? prefs;
```
I would call this preferences

biosfood commented on lib/main.dart:
```diff
@@ -1,20 +1,80 @@
+                                onChanged: (value) {
+                                    setState(() {
+                                        _isDarkMode = value;
+                                         _saveTheme(value);
```
indentation

biosfood commented on lib/main.dart:
```diff
@@ -1,20 +1,80 @@
+                appBar: AppBar(
+                    title: Text('Flutter Theme Switch'),
+                ),
+                drawer: Drawer(
```
Great. I like it

biosfood commented on pubspec.yaml:
```diff
@@ -9,6 +9,7 @@ environment:
 dependencies:
   flutter:
     sdk: flutter
+  shared_preference: ^2.0.13
```
shared_preferences

biosfood replied:
looks good now

biosfood commented on lib/main.dart:
```diff
@@ -1,20 +1,80 @@
+
+void main() async {
+    WidgetsFlutterBinding.ensureInitialized();
+    prefs = await SharedPreferences.getInstance();
```
I would prefer this to happen in the app but flutter seems to handle a loading screen quite nicely so this is okay for now

biosfood reviewed (changes requested):
Organization is good now,
There are still a bunch of compiler warnings and some formatting errors that need fixing

Also, make sure to test everything

biosfood commented:
Also, commits can and should consist of a single "action", you don't need a new commit for every single change.
Doesn't matter now, we can just squash everything in the end

Lemonatix committed:
Update pubspec.yaml

fixes typing mistake

Signed-off-by: Lemonatix <ph.mika.ma.riesterer@gmail.com>

Lemonatix committed:
Reformat and fix problems for dark mode toggle

Signed-off-by: Lemonatix <ph.mika.ma.riesterer@gmail.com>

biosfood committed:
rename prefs to preferences

biosfood reviewed (approved):
Looks good now.
Great job
:)
